crates/runtime/src/memory.rs

```diff
@@ -389,29 +386,33 @@ impl Memory {
     pub unsafe fn grow(
         &mut self,
         delta_pages: u64,
-        limiter: Option<&mut dyn ResourceLimiter>,
+        mut limiter: Option<&mut dyn ResourceLimiter>,
     ) -> Option<usize> {
         let old_byte_size = self.byte_size();
+        // Wasm spec: when growing by 0 pages, always return the current size.
         if delta_pages == 0 {
             return Some(old_byte_size);
         }
 
+        // Use saturating arithmetic to represent any size too big for memory as usize::MAX
         let new_byte_size = usize::try_from(delta_pages)
-            .ok()?
-            .checked_mul(WASM_PAGE_SIZE)?
-            .checked_add(old_byte_size)?;
+            .unwrap_or(usize::MAX)
+            .saturating_mul(WASM_PAGE_SIZE)
+            .saturating_add(old_byte_size);
```

Comment:
I'm a bit wary about saturating here for `new_byte_size`. I think I'd prefer to keep `new_byte_size` as `Option` since it's the precise page-aligned size requested by the wasm, and then when we call the limiter it'd be `new_byte_size.unwrap_or(usize::MAX)` or something like that.

Otherwise this saturated value could leak into `grow_to` below which  runs a risk of getting confused if the size isn't page-aligned.

Reply:
usize::MAX is not page aligned though, its 2^32-1 or 2^64-1. I can change this to checked operations that unwrap to usize::MAX, but grow_to will still need a special case.